src/cmap/connection_pool.ts

```diff
@@ -289,9 +345,19 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
    * Pool reset is handled by incrementing the pool's generation count. Any existing connection of a
    * previous generation will eventually be pruned during subsequent checkouts.
    */
-  clear(): void {
-    this[kGeneration] += 1;
-    this.emit('connectionPoolCleared', new ConnectionPoolClearedEvent(this));
+  clear(serviceId?: ObjectId): void {
+    if (this.loadBalanced && serviceId) {
+      const generation = this.serviceGenerations.get(serviceId);
+      // Only need to worry if the generation exists, since it should
+      // always be there but typescript needs the check.
+      if (generation) {
+        // Increment the generation for the service id.
+        this.serviceGenerations.set(serviceId, generation + 1);
+      }
```

Comment:
what if `generation == null`? Maybe we should throw an error, since that state should be impossible

Reply:
Added throw of `MongoDriverError` in this case.